File: Week 02/id_063/Java/LeetCode_77_063.java
/*
思路

递归算法练习

进行1层到n层的递归，每一层递归选择当前层对应的数字是否选择，当选择是的层数总和达到k时候生成一个结果并且进行回溯
 */



public class Solution {

    private List<List<Integer>> dfs(List<List<Integer>> result, List<Integer> curChoice, int maxChoiceNum, int curLevel, int maxLevel) {
        if (curChoice.size() == maxChoiceNum) {
            result.add(new ArrayList<Integer>(curChoice));
            return result;
        }

        if (curLevel > maxLevel) {
            return result;
        }

        if ((maxLevel - curLevel + 1) + curChoice.size() < maxChoiceNum) {
            // 剪枝操作，就算剩下所有的数字全部选择也凑不齐maxChoiceNum个数字，直接回溯
            return result;
        }

        // 当前层数据不选择
        dfs(result, curChoice, maxChoiceNum, curLevel+1, maxLevel);

        // 当前层数据选择
        curChoice.add(curLevel);
        dfs(result, curChoice, maxChoiceNum, curLevel+1, maxLevel);
        curChoice.remove(curChoice.size() - 1);     // 恢复现场

        return result;
    }

    // 计算组合数
    private int getComposeNum(int n, int k) {
        if (k > n - k) {
            k = n-k;
        }

        int totalNum = 1;
        for (int i = 0; i < k; i++) {
            totalNum *= n - i;
        }

        for (int i = 0; i < k; i++) {
            totalNum /= k;
        }

        return totalNum;
    }

    public List<List<Integer>> combine(int n, int k) {
        return dfs(new ArrayList<List<Integer>>(getComposeNum(n, k)), new ArrayList<Integer>(k), k, 1, n);
    }
}

File: Week 02/id_063/C++/LeetCode_77_063.cpp
/*
思路
递归算法练习
进行1层到n层的递归，每一层递归选择当前层对应的数字是否选择，当选择是的层数总和达到k时候生成一个结果并且进行回溯
*/

#include <vector>
using namespace std;

class Solution {
public:

    void dfs(vector<vector<int>>& result, vector<bool>& choice, int choice_num, int target_choice_num, int cur_level, int max_level) {
        if (cur_level == max_level) {
            return;
        }

        choice[cur_level] = false;
        dfs(result, choice, choice_num, target_choice_num, cur_level + 1, max_level);

        choice[cur_level] = true;
        if (choice_num+1 == target_choice_num) {
            vector<int> v;
            for (int i = 0; i <= cur_level; i++) {
                if (choice[i]) {
                    v.push_back(i+1);
                }
            }
            result.push_back(v);
        } else {
            dfs(result, choice, choice_num + 1, target_choice_num, cur_level + 1, max_level);
        }
    }

    vector<vector<int>> combine(int n, int k) {
        vector<bool> choice(n);
        vector<vector<int>> v;

        if (k == 0) {
            return v;
        }

        dfs(v, choice, 0, k, 0, n);
        return v;
    }
};